### PR TITLE
Add rough graph changes handling by full page reload

### PR DESCRIPTION
**Pull request by samogot**

### CODE REVIEW

File: build_runner/lib/src/server/hot_reload_client/client.dart [hunk 1]
```diff
@@ -128,23 +141,24 @@ List<K> keys<K, V>(JsMap<K, V> map) {
...
-
-Future<Module> _loadModule(String moduleId) {
-  var completer = Completer<Module>();
-  dartLoader.loadModule(
-      moduleId,
-      allowInterop((HotReloadableModule module) =>
-          completer.complete(ModuleWrapper(module))));
-  return completer.future;
-}
+Future<Module> Function(String) _wrapLoaderFunction(
```
Comment (grouma) [inline build_runner/lib/src/server/hot_reload_client/client.dart]: Can you use a more descriptive name here? Something that indicates the purpose of the wrapping.

File: build_runner/lib/src/server/hot_reload_client/reloading_manager.dart [hunk 2]
```diff
@@ -67,49 +67,54 @@ class ReloadingManager {
...
           _reloadPage();
           _running.complete();
           return;
         }
-        _dirtyModules.add(parentId);
+
+        var parentIds = _moduleParents(moduleId);
+        if (parentIds == null || parentIds.isEmpty) {
+          // Propagating from root should cause page reload
+          _reloadPage();
```
Comment (grouma) [inline build_runner/lib/src/server/hot_reload_client/reloading_manager.dart]: Not for this PR but I wonder if we should add some debug level prints for when a hot reload is initiated?

File: build_runner/lib/src/server/hot_reload_client/reloading_manager.dart [hunk 3]
```diff
@@ -67,49 +67,54 @@ class ReloadingManager {
...
+          if (success == true) continue;
+          if (success == false) {
+            _reloadPage();
+            _running.complete();
+            return;
+          }
+          _dirtyModules.add(parentId);
+        }
       }
+    } on DeferredLoadException catch (e) {
```
Comment (jakemac53) [inline build_runner/lib/src/server/hot_reload_client/reloading_manager.dart]: `DeferredLoadException` isn't really the correct type of exception here (even though it sounds sort of right). This is meant for deferred loaded Dart imports.

I would just create your own exception type.

File: build_runner/test/server/hot_reload_client/reloading_manager_test.dart [hunk 4]
```diff
@@ -71,6 +71,13 @@ void main() {
     verifyInOrder([methods.reloadModule('root'), methods.reloadPage()]);
   });
 
+  test('reloadind error triggers page reload', () async {
```
Comment (jakemac53) [inline build_runner/test/server/hot_reload_client/reloading_manager_test.dart]: s/reloadind/reloading

File: build_web_compilers/lib/src/dev_compiler_bootstrap.dart [hunk 5]
```diff
@@ -305,6 +321,20 @@ final _requireJsConfig = '''
 (function() {
   var oldOnError = requirejs.onError;
   requirejs.onError = function(e) {
+    if (e.requireModules) {
+      // If error ocuired on loading dependencies, we need to invalidate ancessor too.
```
Comment (jakemac53) [inline build_web_compilers/lib/src/dev_compiler_bootstrap.dart]: s/ocuired/occurred

File: build_web_compilers/lib/src/dev_compiler_bootstrap.dart [hunk 6]
```diff
@@ -338,12 +368,27 @@ require.config({
...
+      return !previousDeps.includes(dep.name);
+    });
+    if (changed) {
+      console.warn("Dependencies graph change for module '" + map.name + "' detected. " + 
+        "Dependencies was [" + previousDeps + "], now [" +  depArray.map((dep) => dep.name) +"]. " +
+        "Page can't be hot-reloaded, firing full page reload.");
+      window.location.reload();
+    }
+  }
+  \$dartLoader.modulesGraph.set(map.name, []);
```
Comment (jakemac53) [inline build_web_compilers/lib/src/dev_compiler_bootstrap.dart]: should this all move into an else block? I don't think we need to clear it out and set it up again.

**Pull request merged**